File: Graph Theory/leetcode_133.py

# Problem Link == == >> https://leetcode.com/problems/clone-graph/description/

from collections import defaultdict


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Node) -> Node:
        if not node:
            return None

        visited = {}
        return self.dfs(node, visited)

    def dfs(self, node: Node, visited: dict[Node, Node]) -> Node:
        if node in visited:
            return visited[node]

        clone = Node(node.val)
        visited[node] = clone

        for neighbor in node.neighbors:
            clone.neighbors.append(self.dfs(neighbor, visited))

        return clone


# Sample usage in main function
if __name__ == "__main__":
    node1 = Node(1)
    node2 = Node(2)
    node3 = Node(3)

    node1.neighbors.append(node2)
    node1.neighbors.append(node3)
    node2.neighbors.append(node1)
    node2.neighbors.append(node3)
    node3.neighbors.append(node1)
    node3.neighbors.append(node2)

    solution = Solution()
    cloned_graph = solution.cloneGraph(node1)

    # You can now use the cloned_graph here
    print(f"Cloned graph node1 value: {cloned_graph.val}")  # Example usage

File: Graph Theory/leetcode_133.cpp
// Problem Link == == >> https://leetcode.com/problems/clone-graph/description/

#include<bits/stdc++.h>
                              using namespace std;

// Definition for a Node.
class Node
{
public:
    int val;
    vector<Node *> neighbors;
    Node()
    {
        val = 0;
        neighbors = vector<Node *>();
    }
    Node(int _val)
    {
        val = _val;
        neighbors = vector<Node *>();
    }
    Node(int _val, vector<Node *> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution
{
public:
    Node *cloneGraph(Node *node)
    {
        if (node == nullptr)
        {
            return nullptr;
        }

        unordered_map<Node *, Node *> visited;
        return dfs(node, visited);
    }

private:
    Node *dfs(Node *node, unordered_map<Node *, Node *> &visited)
    {
        if (visited.count(node))
        {
            return visited[node];
        }

        Node *clone = new Node(node->val);
        visited[node] = clone;

        for (Node *neighbor : node->neighbors)
        {
            clone->neighbors.push_back(dfs(neighbor, visited));
        }

        return clone;
    }
};

int main()
{
    // Create a sample graph
    Node *node1 = new Node(1);
    Node *node2 = new Node(2);
    Node *node3 = new Node(3);

    node1->neighbors.push_back(node2);
    node1->neighbors.push_back(node3);
    node2->neighbors.push_back(node1);
    node2->neighbors.push_back(node3);
    node3->neighbors.push_back(node1);
    node3->neighbors.push_back(node2);

    // Clone the graph
    Solution solution;
    Node *clonedGraph = solution.cloneGraph(node1);

    // You can now use the cloned graph here

    return 0;
}
